### DISCUSSION TITLE
Reusable Window with dynamic window content

### DISCUSSION THREAD

**Discussion by TheColorRed** [slint-ui/slint#6772] [Q&A]
I am making a project that will use a lot of windows. Is there a way that I can easy use a base window component and dynamically load the content without having to do an if statement for each component type that it should load?

Currently the window looks something like this, where the top bar will be static on all windows, but the content will change for every window.

![image](https://github.com/user-attachments/assets/60ed980a-1bd1-4102-93ae-5ac7c8f58633)

In the rectangle section, to get this to work I will have to do an if statement for every window which would will start to get redundant especially for the amount of windows I plan on having.

```slint
export component BaseWindow inherits Window {
    no-frame: true;
    in-out property <string> window-title <=> title-bar.title;

    in property <string> content: "new-project";

    VerticalLayout {
        title-bar := TitleBar {
            width: 100%;
            has-menu-bar: false;
        }

        Rectangle {
            if(content == "new-project"): NewProjectDialog {
                init => {
                    root.window-title = "New Project";
                    root.width = 800px;
                    root.height = 600px;
                }
            }
        }
    }
}
```

Is there maybe a way to dynamically load files, maybe something like this:

```slint
        Rectangle {
            @import("./window-content/" + content + ".slint");
        }
```

or even through the rust code (pseudo code) where maybe we can have a `component` type (which I don't believe exists as a type).

```rs
slint::slint! {
  export component BaseWindow inherits Window {
    in property <component> my-component;
    VerticalLayout {
      ComponentRenderer { component: my-component; }
    }
  }
}

fn main() {
  let app = AppWindow::new().unwrap();

  app.on_open_window(||{
    let base = BaseWindow::new().unwrap();
    // Where `my-component` is of type `<component>`.
    base.set_my_component(NewProjectDialog);
  });
}
```

**Comment by hunger** [comment]
The `if` approach is the most stable one right now.

We do have an experimental `ComponentFactory` though, which is what we use to embed an interpreter into the live preview UI. You could try that -- it is rust only right now though and not yet polished. That could be used to dynamically load a file into a UI.
